examples/plot_scalable_poly_kernels.py

```diff
@@ -0,0 +1,187 @@
+from sklearn.svm import LinearSVC
+from sklearn.kernel_approximation import PolynomialSampler
+from sklearn.pipeline import Pipeline
+import time
+
+##############################################################################
+# Load the Covtype dataset, which contains 581,012 samples
+# with 54 features each, distributed among 6 classes. The goal of this dataset
+# is to predict forest cover type from cartographic variables only
+# (no remotely sensed data). After loading, we transform it into a binary
+# classification problem to match the version of the dataset in the
+# LIBSVM webpage [2], which was the one used in [1].
+
+data = fetch_covtype()
+X, y = data["data"], data["target"]
+
+y[y != 2] = 0
+y[y == 2] = 1  # We will try to separate class 2 from the other 6 classes.
+
+##############################################################################
+# Here we select 5,000 samples for training and 10,000 for testing.
+# To actually reproduce the results in the original Tensor Sketch paper,
+# select 100,000 for training.
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=5_000,
+                                                    test_size=10_000,
+                                                    random_state=42)
+
+##############################################################################
+# Now scale features to the range [0, 1] to match the format of the dataset in
+# the LIBSVM webpage, and then normalize to unit length as done in the
+# original Tensor Sketch paper [1].
+
+mm = MinMaxScaler().fit(X_train)
+X_train = normalize(mm.transform(X_train))
+X_test = normalize(mm.transform(X_test))
+
+
+##############################################################################
+# As a baseline, train a linear SVM on the original features and print the
+# accuracy. We also measure and store accuracies and training times to
+# plot them latter.
+
+results = {}
+
+lsvm = LinearSVC()
+start = time.time()
+lsvm.fit(X_train, y_train)
+lsvm_time = time.time() - start
+lsvm_score = 100 * lsvm.score(X_test, y_test)
+
+results[f"LSVM"] = {"time": lsvm_time, "score": lsvm_score}
+print(f"Linear SVM score on raw features: {lsvm_score:.2f}%")
+
+##############################################################################
+# Then we train linear SVMs on the features generated by
+# :class:`PolynomialSampler` with different values for `n_components`, showing
+# that these kernel feature approximations improve the accuracy of linear
+# classification. Repeat the experiment 5 times to compensate for the
+# stochastic nature of :class:`PolynomialSampler`.
+
+n_runs = 5
+for n_components in [250, 500, 1000, 2000]:
+
+    ps_lsvm_time = 0
+    ps_lsvm_score = 0
+    for _ in range(n_runs):
+
+        pipeline = Pipeline(steps=[("kernel_approximator",
+                                    PolynomialSampler(
+                                        n_components=n_components,
+                                        degree=4)),
```

Comment:
Can you say something about the `degree` in the example. With degree 4, there would be 54^4 ~ 8E6 output features. That's where `n_components` comes in.

Reply:
done!

Reply:
Sorry, I mixed up the polynomial kernel with `preprocessing.PolynomialFeatures`. It does not produce 54^4 output features.

Reply:
yeah, sure not! But the explicit kernel feature space would have indeed 54^4 features. What I added was:

> Note that, since the original samples have 54 features, the explicit feature map of the polynomial kernel of degree four would have 54^4 features. Thanks to :class:`PolynomialSampler`, we can condense most of the discriminative information of that feature space into a much more compact representation.

So the thing is that, to access the power of the kernel, you can 1) use a kernelized SVM, which uses the kernel function to access the feature space in an implicit manner, but has poor scalability. 2) Use the explicit kernel feature space, which in this case is unfeasible because it would have 54^4 features, or 3) Use a kernel approximation method such as PolynomialSampler, to condense the information of the kernel feature space to a compact representation.

Let me know if you would like to remove the note I added! It is accurate, but perhaps to the best place to mention this.

Reply:
Thanks for the explanation! So I was not so wrong after all:smile:
In this case, it is good to mention it and also give a number 54^4 ~ xxx. This way, the `n_components` becomes a better meaning.